while-loops/problem-set-while-loops.md

```diff
@@ -0,0 +1,849 @@
+
+##### !end-explanation
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+<!-- prettier-ignore-end -->
+
+
+<!-- prettier-ignore-start -->
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: code-snippet
+* language: python3.6
+* id: 85dffbd3-2030-44b4-8b9f-cc86f0c7e9d1
+* title: While Loops
+
+##### !question
+
+Let's implement a function called `sidewinder`. It will take a parameter `number_list` which contains a list of the _next_ index in the list to visit. Starting from position `0`, we will look in the current position, get the value stored there, and use it as the _next_ position to visit. We will continue doing this until we find a value of `0`, which means we have found the end of the index chain. Then we'll return the position where we found the `0`.
+
+For example, if we received `[2, 0, 1]` as our `number_list`, starting from position `0` we find a `2`. We look in position `2` and find a `1`. So we look in position `1` and find `0`. `0` marks the end of the chain, so we return `1`, the position we just looked in.
+
+The diagram below shows a longer example. Notice that there are _multiple_ zeros in the list, so we _cannot_ iterate through in order and return the position of the first zero that we find.
+
+<div style="max-width: 700px; margin: 0 auto -28px auto; text-align: center;">
+<img alt="How to interpret the array input to sidewinder" src="../assets/while-loops_problem-set-while-loops_sidewinder.png">
+</div>
+*Fig. Given the list `[4, 9, 0, 0, 1, 8, 0, 0, 11, 5, 0, 10]` for `number_list`, this diagram shows how to move through the list values to find the 'end.' Starting from index `0`, we find the value `4`. We go to index `4` where we find the value `1`. We go to index `1` where we find the value `9`, and so on. Eventually we reach index `10` where we find the value `0`, so this is the end and the result is `10`.*
+
+Provide an implementation for `sidewinder` so that it
+- follows the chain of indices to the end, and
+- returns the position of the end of the chain
+
+You may assume that `number_list` will never be an empty list, and that all of the values in the list are valid indices in the list.
+
+##### !end-question
+
+##### !placeholder
+
+```python
+def sidewinder(number_list):
+    pass
+```
+
+##### !end-placeholder
+
+##### !tests
+
+```python
+import unittest
+from main import sidewinder
+
+class TestPython1(unittest.TestCase):
+  def test_one(self):
+    self.assertEqual(0, sidewinder([0]))
+
+  def test_two(self):
+    self.assertEqual(1, sidewinder([2, 0, 1]))
+
+  def test_three(self):
```

Comment:
Students can see these test names when debugging, so it's helpful to rename them

```suggestion
  def test_sidewinder_when_list_of_one_element_zero(self):
    self.assertEqual(0, sidewinder([0]))

  def test_sidewinder_stops_at_value_zero(self):
    self.assertEqual(1, sidewinder([2, 0, 1]))

  def test_stops_at_value_zero_in_long_list(self):
```